ASSIGNMENT 9

1. Load HR-Employee-Attrition.csv dataset and create an ensemble ML model for predicting target variable (Attrition). Report the performance of the model using appropriate metrics.

In [2]:
#Import the dataset to google colab
import pandas as pd
from google.colab import files
import io
uploaded = files.upload()                                                    #upload files 
data = pd.read_csv(io.BytesIO(uploaded['HR-Employee-Attrition.csv']))        #read the dataset

Saving HR-Employee-Attrition.csv to HR-Employee-Attrition (7).csv


In [3]:
data.dtypes

Age                          int64
Attrition                   object
BusinessTravel              object
DailyRate                    int64
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
EmployeeCount                int64
EmployeeNumber               int64
EnvironmentSatisfaction      int64
Gender                      object
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
MaritalStatus               object
MonthlyIncome                int64
MonthlyRate                  int64
NumCompaniesWorked           int64
Over18                      object
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StandardHours                int64
StockOptionLevel             int64
TotalWorkingYears   

In [4]:
data['Department'] = data.Department.astype(bool)                                           
data['Gender'] = data.Gender.astype(bool) 
data['OverTime'] = data.OverTime.astype(bool)     
data['Attrition'] = data.Attrition.astype(bool)  
data.dtypes                                                                                 #Print the datatypes

Age                          int64
Attrition                     bool
BusinessTravel              object
DailyRate                    int64
Department                    bool
DistanceFromHome             int64
Education                    int64
EducationField              object
EmployeeCount                int64
EmployeeNumber               int64
EnvironmentSatisfaction      int64
Gender                        bool
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
MaritalStatus               object
MonthlyIncome                int64
MonthlyRate                  int64
NumCompaniesWorked           int64
Over18                      object
OverTime                      bool
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StandardHours                int64
StockOptionLevel             int64
TotalWorkingYears   

In [5]:
#####Convert the columns of the dataset######
data['Department']= data['Department'].map({'Sales': '1', 'Research & Development': '0'})        #Convert response column
data['Gender'] = data['Gender'].map({'Male': '1', 'Female': '0'})              
data['OverTime'] = data['OverTime'].map({'Yes': '1', 'No': '0'}) 
data['Attrition'] = data['Attrition'].map({'Yes': '1', 'No': '0'})                                                          

In [8]:
from sklearn.preprocessing import OneHotEncoder      
enc = OneHotEncoder()                                
data = enc.fit_transform(data).toarray()    

In [9]:
X = data[:, [0,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34]]           
Y = data[:, 1]                                                           

In [10]:
from sklearn.model_selection import train_test_split                         
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.20)    

In [21]:
import numpy as np
from sklearn.metrics import  accuracy_score, zero_one_loss,precision_score,confusion_matrix, recall_score, f1_score
from sklearn.ensemble import AdaBoostClassifier 
model = AdaBoostClassifier(n_estimators=10,learning_rate=0.1 ,random_state=42)             
model.fit(X_train, Y_train)                                                           
predict = model.predict(X_test) 
print('Accuracy:',round(accuracy_score(Y_test, predict),4)*100)                                #Calculate Accuracy score
print('Error:',round(zero_one_loss(Y_test, predict),4)*100)                                    #Calculate Error Score
print('F-score:',round(f1_score(Y_test,predict,average='weighted'),4)*100)                     #Calculate F_score value
print('Recall:', round(recall_score(Y_test, predict, average='weighted'),4)*100)               #Calculate Recall value

Accuracy: 99.66000000000001
Error: 0.33999999999999997
F-score: 99.49
Recall: 99.66000000000001


2. Have you used any hyperparameter tuning while building the model in Q1? If so then plot your performance metrics for different hyperparmeter values that you have used in Q1.  
I have not used hyper parameter tuning in Q1. 

3. Reflect on the importance of hyperparameter tuning of ML models based on your ML model development exercise. 
Hyperparameter tuning is important factor while designing an Machine learning model. The parameters decide the model performance and accuracy rate. I have used different hyperparameter modelling technique ion my previous task which shows that model performance is directly proportional to parameter tuning. 

4. Create a GradientBoost model for the predicting Attrition using the same dataset that you have used Q1 and report the performance.

In [24]:
x = data[:, [0,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34]]           
y = data[:, 1]                                                           

In [25]:
from sklearn.model_selection import train_test_split                         
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.20)    

In [26]:
#Gradient Boosting model
from xgboost import XGBClassifier 
grad = XGBClassifier(learning_rate =1,max_depth=2,objective= 'binary:logistic',nthread=2)
grad.fit(x_train, y_train)  
pred_y = grad.predict(x_test)
print('Accuracy:',round(accuracy_score(y_test, pred_y),4)*100)                                #Calculate Accuracy score
print('Error:',round(zero_one_loss(y_test, pred_y),4)*100)                                    #Calculate Error Score
print('F-score:',round(f1_score(y_test,pred_y,average='weighted'),4)*100)                     #Calculate F_score value
print('Recall:', round(recall_score(y_test, pred_y, average='weighted'),4)*100)               #Calculate Recall value


Accuracy: 98.3
Error: 1.7000000000000002
F-score: 97.46000000000001
Recall: 98.3


5. Compare the performance of two models (Q1 and Q3). Explain which model is good and why.

The Adaboost Boost classifier has better model performance as compared to Gradient boost classifier. The model accuracy rate of Adaboost classifier is 99.6% whereas the accuracy rate of Gradient boost classifier is 98.3%.
Adaboost classifier minimizes the loss function making the model performance better. 